drivers/console/aarch64/console.S

```diff
@@ -3,103 +3,240 @@
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
-#include <asm_macros.S>
 
-	.globl	console_init
+#include <console_macros.S>
+
+	.globl	console_register
+	.globl	console_unregister
 	.globl	console_uninit
+	.globl	console_start_crash
 	.globl	console_putc
 	.globl	console_getc
 	.globl	console_flush
 
 	/*
-	 *  The console base is in the data section and not in .bss
-	 *  even though it is zero-init. In particular, this allows
+	 *  The console list pointer is in the data section and not in
+	 *  .bss even though it is zero-init. In particular, this allows
 	 *  the console functions to start using this variable before
 	 *  the runtime memory is initialized for images which do not
 	 *  need to copy the .data section from ROM to RAM.
 	 */
-.section .data.console_base ; .align 3
-	console_base: .quad 0x0
+.section .data.console_list ; .align 3
+	console_list: .quad 0x0
+.section .data.console_state ; .align 0
+	console_state: .byte (1 << CONSOLE_FLAG_BOOT_SHIFT)
 
 	/* -----------------------------------------------
+	 * void console_register(console_info_t *console)
+	 * Function to insert a new console structure into
+	 * the console list. Should usually only be called
+	 * by declare_console expansions.
+	 * In : x0 - address of console structure
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_register
+	adrp	x14, console_list
+	ldr	x1, [x14, :lo12:console_list]	/* X1 = first struct in list */
+	str	x0, [x14, :lo12:console_list]	/* list head = new console */
+	str	x1, [x0]			/* new console next ptr = X1 */
+	ret
+endfunc console_register
+
+#if !ERROR_DEPRECATED
+	/* Set up compatiblity symbols for platforms using old console API. */
+	declare_console core
+
+	.globl	console_init
+	/* -----------------------------------------------
 	 * int console_init(uintptr_t base_addr,
 	 * unsigned int uart_clk, unsigned int baud_rate)
-	 * Function to initialize the console without a
-	 * C Runtime to print debug information. It saves
-	 * the console base to the data section.
+	 * Deprecated legacy function for initializing a
+	 * single console. Current platforms should call
+	 * console_<driver>_register for the relevant
+	 * driver instead. This needs to clear the console
+	 * list first because some legacy platforms called
+	 * console_init more than once and expect it to
+	 * overwrite previous console state.
```

Comment:
OK, I agree console_init() should be deprecated. Ideally the declaration in the header should add the __deprecated keyword but we can only do that if AArch32 is updated too.

Regarding multiple instances of the same driver, I agree its unlikely this will be needed (for now). However, if this were needed in future, it would be possible to create a simple memory pool rather than full dynamic memory allocation. Anyway, let's ignore this for now.